API Replace Page Delete with new Page Archive action

tractorcow opened the pull request:
The major distinction between archiving and deleting is that "deleting" only deletes from stage (which is practically pretty useless), but archiving removes both live and stage, but informs the user that the page is safely recorded in the archive.

"delete" gives the false impression that the page is lost forever, which has never been the case.

This PR must be merged and reviewed alongside a CMS pull request incoming.

Lots of old code has been updated and fixed; Some which hasn't worked since 3.0 (e.g. batch action confirmations).

![image](https://cloud.githubusercontent.com/assets/936064/7647228/2bfa2ea4-fb25-11e4-938c-b302dae9a4ad.png)

tractorcow commented:
Ok the CMS pr is at https://github.com/silverstripe/silverstripe-cms/pull/1203

tractorcow commented:
This is a semi-major change, so please ensure that this PR is left open with plenty of time for users to review and give feedback.

tractorcow commented on admin/javascript/LeftAndMain.BatchActions.js:
```diff
@@ -38,12 +38,89 @@
+			 */
+			registerDefault: function() {
+				// Publish selected pages action
+				this.register('admin/pages/batchactions/publish', function(ids) {
```
I found this code here, and it hasn't been working for years. ;)

assertchris commented:
/me stops hovering over the merge button...

tractorcow commented on admin/javascript/LeftAndMain.BatchActions.js:
```diff
@@ -319,8 +420,10 @@
+				}
+				
+				// Refresh selected / enabled nodes
+				$('#Form_BatchActionsForm').refreshSelected();
```
This code was removed years ago; Re-added this in to ensure tree items are properly filtered.

tractorcow commented:
Update to this; I've re-tested and fixed some of the CSS rules in the sitetree, and fixed some issues with recent work to make it work with the updated flags / styles.

tractorcow commented:
A few updates:

- Restoring an archived pages now restores its parents
- Some more CSS cleanup / fixes to make sure that disabled / unfiltered styles are properly inherited

chillu commented:
So excited this is finally happening, great work guys! I've just checked it out locally. Long-term I think we need to make it more obvious on how to find archived pages in the site tree again - at the moment they stay in the tree until you reload, which might come as a surprise to users. You can only get them back through the filter panel.

On that note: You should update the userhelp ;)

tractorcow commented:
I've updated the other PR (https://github.com/silverstripe/silverstripe-cms/pull/1203) based on feedback:

Instead of automatically restoring the parent page, the user is given the option to restore any orphaned pages to the root. The text / description / alert on the action will change depending on the parent page.

tractorcow commented:
More bugfixes to https://github.com/silverstripe/silverstripe-cms/pull/1203:

- Side by side preview was not working on archive action, so ContentController has been fixed for checking permissions of the page in the current stage
- Fixed some typos in the new alerts
- Changed "Un-publish" to "Unpublish"
- Changed "Restore" and "Restore at top level" to "Restore draft" and "Restore draft at top level"

tractorcow commented:
Rebased with new colours / hover behaviours. :D

![image](https://cloud.githubusercontent.com/assets/936064/7760496/4c5564c0-0070-11e5-8f86-ceba141f1dc9.png)

![image](https://cloud.githubusercontent.com/assets/936064/7760498/5a6e95ea-0070-11e5-94c4-108d965062fe.png)

tractorcow commented:
Rebased again

sminnee commented:
This looks nice, although I think that `applicablePages()` might need a test, and I'd also be keen to see how this interacts with advancedworkflow, as a smoke-test of how complicated modules might be broken by this change.

tractorcow commented:
There aren't any tests anywhere for batch actions... so... I'll just have to make all new ones I guess.

tractorcow commented:
@sminnee I've added tests but they are on the https://github.com/silverstripe/silverstripe-cms/pull/1203 PR, since framework doesn't actually implement that API anywhere.

tractorcow commented:
Rebased again

sminnee commented:
It seems that this breaks the build on postgres, although I can't understand why. :-(

assertchris commented:
https://github.com/silverstripe/silverstripe-framework/pull/4189/files#diff-69e8356e095b9501b3b000dbed87a13dR276 maybe?

tractorcow commented:
Nothing to do with postgres; It's that one test that only sometimes passes. I restarted the build for postgres.

tractorcow commented:
Fixed :D

tractorcow commented:
Yay thanks

chillu commented:
Related: https://github.com/silverstripe/silverstripe-framework/issues/6202
